cmd/signing.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return fmt.Errorf(
+			"failed to write to file [%s]: [%v]",
+			targetFilePath,
+			err,
+		)
+	}
+
+	logger.Infof(
+		"key share has been decrypted successfully and written to file [%s]",
+		targetFilePath,
+	)
+
+	return nil
+}
+
+// SignDigest signs a given digest using key shares from the provided directory.
+func SignDigest(c *cli.Context) error {
+	digest := c.String("digest")
+	if len(digest) == 0 {
+		return fmt.Errorf("invalid digest")
+	}
+
+	keySharesDir := c.String("key-shares-dir")
+	if len(keySharesDir) == 0 {
+		return fmt.Errorf("invalid key shares directory name")
+	}
+
+	keySharesFiles, err := ioutil.ReadDir(keySharesDir)
+	if err != nil {
+		return fmt.Errorf(
+			"could not read key shares directory: [%v]",
+			err,
+		)
+	}
+
+	signers := make([]tss.ThresholdSigner, len(keySharesFiles))
+	networkProviders := make([]net.Provider, len(keySharesFiles))
+
+	for i, keyShareFile := range keySharesFiles {
+		keyShareBytes, err := ioutil.ReadFile(
+			fmt.Sprintf("%s/%s", keySharesDir, keyShareFile.Name()),
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"could not read key share file [%v]: [%v]",
+				keyShareFile.Name(),
+				err,
+			)
+		}
+
+		var signer tss.ThresholdSigner
+		err = signer.Unmarshal(keyShareBytes)
+		if err != nil {
+			return fmt.Errorf(
+				"could not unmarshal signer from file [%v]: [%v]",
+				keyShareFile.Name(),
+				err,
+			)
+		}
+
+		operatorPublicKey, err := signer.MemberID().PublicKey()
+		if err != nil {
+			return fmt.Errorf(
+				"could not get operator public key: [%v]",
+				err,
+			)
+		}
+
+		networkKey := key.NetworkPublic(*operatorPublicKey)
+		networkProvider := local.ConnectWithKey(&networkKey)
+
+		signers[i] = signer
+		networkProviders[i] = networkProvider
+	}
+
+	digestBytes, err := hex.DecodeString(digest)
+	if err != nil {
+		return fmt.Errorf("could not decode digest string: [%v]", err)
+	}
+
+	ctx, cancelCtx := context.WithTimeout(
+		context.Background(),
+		1*time.Minute,
+	)
+	defer cancelCtx()
+
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(len(signers))
+
+	type signingOutcome struct {
+		signerIndex int
+		signature   *ecdsa.Signature
+		err         error
+	}
+
+	signingOutcomesChannel := make(chan *signingOutcome, len(signers))
+
+	for i := range signers {
+		go func(signerIndex int) {
+			defer waitGroup.Done()
+
+			signature, err := signers[signerIndex].CalculateSignature(
+				ctx,
+				digestBytes,
+				networkProviders[signerIndex],
+			)
+
+			signingOutcomesChannel <- &signingOutcome{
+				signerIndex,
+				signature,
+				err,
+			}
+		}(i)
+	}
+
+	waitGroup.Wait()
+	close(signingOutcomesChannel)
+
+	for signingOutcome := range signingOutcomesChannel {
+		if signingOutcome.err != nil {
+			logger.Errorf(
+				"[signer:%v] error: [%v]",
+				signingOutcome.signerIndex,
+				signingOutcome.err,
+			)
+			continue
+		}
+
+		logger.Infof(
+			"[signer:%v] signature: [%+v]",
```

Comment:
Is the outcome here a single signature? I don't want to slow us down too much, but I would expect this to:
- Gather all available signatures.
- If there are <`len(signers)`, print an error to stderr and `exit 1`.
- If there are `len(signers)` signatures and they are not all the same, print an error to stderr (possibly listing the three signatures by signer id) and `exit 1`.
- If there are `len(signers)` signatures and they are all the same, print the signature to `stdout` and exit normally (which is to say, return).

I think the error to stderr + exit 1 is doable by just `return`ing an `error`.